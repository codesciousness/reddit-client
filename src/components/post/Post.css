.Post__trending {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 200px;
    height: 200px;
    background: white;
    border-radius: 20px;
    box-shadow: 5px 5px 5px rgba(145, 145, 165, 0.6);
    position: relative;
    margin: 2rem;
    cursor: pointer;
    transition: all 0.3s linear;
}

.Post__trending:hover {
    transform: scale(1.2);
}

.Post__trending::before {
    content: '';
    width: 100%;
    height: 100%;
    background: linear-gradient(rgba(0, 0, 0, 0.5), rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.9));
    border-radius: 20px;
    position: absolute;
    z-index: 1;
    
}

.Post__trending:hover::before {
    background: linear-gradient(rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.5));
}

.Post__trending__image {
    width: 100%;
    height: auto;
    border-radius: 20px;
    position: absolute;
}

.Post__trending__content {
    padding: 1rem;
    z-index: 2;
}

.Post__trending__flair {
    display: inline-block;
    font-size: 1rem;
    font-weight: bold;
    text-transform: uppercase;
    background: lightgray;
    border-radius: 5px;
    margin-bottom: 1rem;
    padding: 0.25rem 0.5rem;
}

.Post__trending__title {
    color: white;
    font-size: 1.25rem;
}

.Post__latest {
    display: flex;
    justify-content: center;
    align-items: center;
    background: rgba(255, 255, 255, 0.7);
    border-radius: 20px;
    box-shadow: 0 5px 10px rgba(145, 145, 165, 0.6);
    margin: 2rem auto;
    position: relative;
    transition: all 0.3s ease-in-out;
}

.Post__latest:hover {
    background: white;
    border-radius: 0;
    border-top: 3px solid gray;
    border-bottom: 3px solid gray;
    border-image: linear-gradient(to right, rgb(255, 2, 108), orange, yellow, green, blue, purple) 1;
}

.Post__latest__image {
    height: 140px;
    width: 140px;
    border-radius: 20px;
    box-shadow: 5px 5px 10px rgba(145, 145, 165, 0.6);
    position: absolute;
    left: -64px;
    cursor: pointer;
    transition: all 0.3s linear 0.1s;
}

.Post__latest__image:hover {
    left: -70px;
}

.Post__latest__content {
    width: 75%;
    padding: 1rem;
}

.Post__latest__flair {
    display: inline-block;
    font-weight: bold;
    text-transform: uppercase;
    background: lightgray;
    border-radius: 5px;
    padding: 0.25rem 0.5rem;
    margin: 0.5rem;
}

.Post__latest__title {
    font-size: 1.5rem;
    margin-bottom: 1rem;
    cursor: pointer;
}

.Post__latest__button {
    font-family: 'Asap', sans-serif;
    font-weight: bold;
    text-align: center;
    text-transform: uppercase;
    color: white;
    background: linear-gradient(to right, rgb(255, 2, 108), orange, yellow, green, blue, purple);
    background-size: 300% 100%;
    border-radius: 50px;
    box-shadow: 0 5px 10px rgba(242, 97, 170, 0.6);
    margin: 0 0.5rem 0.5rem 0.5rem;
    padding: 0.75rem 1.5rem;
    cursor: pointer;
    transition: all 0.4s ease-in-out;
}

.Post__latest__button:hover {
    background-position: 100% 0;
}

.Post__latest__button:hover .Post__latest__icon {
    display: inline;
}

.Post__latest__icon {
    display: none;
}

.Post__latest__detail {
    margin-top: 0.5rem;
}

.Post__latest__detail span {
    margin: 0.5rem 2rem;
}

.Post__latest__author strong {
    color: steelblue;
}

.Post__latest__comments {
    cursor: pointer;
}

.Post__latest__comments:hover {
    color: steelblue;
}