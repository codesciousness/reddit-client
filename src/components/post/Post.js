import React from 'react';
import { useDispatch } from 'react-redux';
import './Post.css';
import PostTime from '../postTime/PostTime';
import { setCurrentPost } from '../../features/posts/postsSlice';

const Post = ({ post, isTrending }) => {
    const dispatch = useDispatch();
    const flair = post.flair.split(' ')[0].toLowerCase();
    const words = post.title.split(' ');
    let shortTitle = [];
    for (let i = 0; i <= words.length; i++) {
        if (words[i] && i < 7) {
            shortTitle.push(words[i]);
        }
    }
    shortTitle = shortTitle.join(' ');
    shortTitle += '...';

    const handleClick = ({ target }) => {
        dispatch(setCurrentPost(target.id));
    };

    if (isTrending) {
        return (
            <section id={post.id} className="Post__trending" onClick={handleClick}>
                <img className="Post__trending__image" src={`https://source.unsplash.com/140x140/?${flair}&sig=${post.id}`} alt=""/>
                <div className="Post__trending__content">
                    <h2 className="Post__trending__flair">{post.flair}</h2>
                    <h3 className="Post__trending__title">{shortTitle}</h3>
                </div> 
            </section>
        );
    }
    else {
        return (
            <section id={post.id} className="Post__latest" onClick={handleClick}>
                <img className="Post__latest__image" src={`https://source.unsplash.com/140x140/?${flair}&sig=${post.id}`} alt=""/>
                <div className="Post__latest__content">
                    <p className="Post__latest__flair">{post.flair}</p>
                    <h2 className="Post__latest__title">{post.title}</h2>
                    <a href={post.url} target="_blank" rel="noreferrer"><button className="Post__latest__button">Read more <i class="fas fa-angle-double-right Post__latest__icon"></i></button></a>
                    <p className="Post__latest__detail">
                        <span className="Post__latest__author">Posted by <strong>{post.author}</strong></span>
                        <PostTime time={post.created} />
                        <span className="Post__latest__comments"><i className="far fa-comment-alt"></i> {post.commentNum}</span>
                    </p>
                </div>
            </section>
        );
    }
}

export default Post;