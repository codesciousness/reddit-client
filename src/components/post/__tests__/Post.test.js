import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { create } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Post from '../Post';

const post = {
    id: 'efgh2i',
    title: 'Where Can the Solar Eclipse be Viewed Next Week',
    author: 'ScienceResearcher',
    created: '30 minutes ago',
    flair: 'Astronomy',
    thumbnail: 'https://source.unsplash.com/140x140/?astronomy',
    commentNum: 25,
    url: 'https://www.sciencemag.org/'
};

jest.mock('../../postTime/PostTime');

const mockStore = configureStore([]);

describe('<Post />', () => {
    let store;

    beforeEach(() => {
        store = mockStore({});
        //Jest Spy Method: store.dispatch = jest.fn();
    });

    test('Post component is defined', () => {
        expect(<Post />).toBeDefined();
    });
    test('matches the trending post snapshot', () => {
        const component = create(
            <Provider store={store}>
                <Router>
                    <Post post={post} isTrending={true} />
                </Router>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('matches the latest post snapshot', () => {
        const component = create(
            <Provider store={store}>
                <Router>
                    <Post post={post} isTrending={false} />
                </Router>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('renders correctly when isTrending=true prop is passed', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Post post={post} isTrending={true} />
                </Router>
            </Provider>
        );
        const title = container.querySelector('.Post__trending__title').textContent;
        const linkComponent = container.querySelector('.Post__trending__link');
        const titleLength = title.split(' ').length
        expect(titleLength).toBeLessThanOrEqual(7);
        expect(screen.getByText('Astronomy')).toBeInTheDocument();
        expect(linkComponent).toBeInTheDocument();
        expect(screen.queryByRole('button')).not.toBeInTheDocument();
        expect(screen.queryByText('ScienceResearcher')).not.toBeInTheDocument();
        expect(screen.queryByTestId('PostTime')).not.toBeInTheDocument();
    });
    test('renders correctly when isTrending=false prop is passed', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Post post={post} isTrending={false} />
                </Router>
            </Provider>
        );
        const title = container.querySelector('.Post__latest__title');
        const linkComponent = container.querySelector('.Post__latest__link')
        expect(title).toBeInTheDocument();
        expect(screen.getByText('Astronomy')).toBeInTheDocument();
        expect(linkComponent).toBeInTheDocument();
        expect(screen.getByRole('button')).toBeInTheDocument();
        expect(screen.getByText('ScienceResearcher')).toBeInTheDocument();
        expect(screen.getByTestId('PostTime')).toBeInTheDocument();
    });
    test('dispatches correct action when trending post title is clicked', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Post post={post} isTrending={true} />
                </Router>
            </Provider>
        );
        const title = container.querySelector('.Post__trending__title');
        userEvent.click(title);
        const actions = store.getActions();
        const expectedPayload = {type: 'posts/setCurrentPost', payload: 'efgh2i'};
        expect(actions).toEqual([expectedPayload]);
        /* Jest Spy Method:
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith({type: 'posts/setCurrentPost', payload: 'efgh2i'});
        */
    });
    test('dispatches correct action when latest post title is clicked', () => {
        const { container } = render(
            <Provider store={store}>
                <Router>
                    <Post post={post} isTrending={false} />
                </Router>
            </Provider>
        );
        const title = container.querySelector('.Post__latest__title');
        userEvent.click(title);
        const actions = store.getActions();
        const expectedPayload = {type: 'posts/setCurrentPost', payload: 'efgh2i'};
        expect(actions).toEqual([expectedPayload]);
        /* Jest Spy Method:
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith({type: 'posts/setCurrentPost', payload: 'efgh2i'});
        */
    });
});