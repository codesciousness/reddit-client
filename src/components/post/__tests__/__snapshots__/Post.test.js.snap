// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post /> matches the latest post snapshot 1`] = `
<section
  className="Post__latest"
>
  <img
    alt=""
    className="Post__latest__image"
    src="https://source.unsplash.com/140x140/?astronomy&sig=efgh2i"
  />
  <div
    className="Post__latest__content"
  >
    <p
      className="Post__latest__flair"
    >
      Astronomy
    </p>
    <a
      className="Post__latest__link"
      href="/posts/astronomy/efgh2i"
      onClick={[Function]}
    >
      <h2
        className="Post__latest__title"
        id="efgh2i"
        onClick={[Function]}
      >
        Where Can the Solar Eclipse be Viewed Next Week
      </h2>
    </a>
    <a
      href="https://www.sciencemag.org/"
      rel="noreferrer"
      target="_blank"
    >
      <button
        className="Post__latest__button"
      >
        Read more 
        <i
          className="fas fa-angle-double-right Post__latest__icon"
        />
      </button>
    </a>
    <p
      className="Post__latest__detail"
    >
      <span
        className="Post__latest__author"
      >
        Posted by 
        <strong>
          ScienceResearcher
        </strong>
      </span>
      <span
        data-testid="PostTime"
      >
        30 minutes ago
      </span>
      <span
        className="Post__latest__comments"
      >
        <i
          className="far fa-comment-alt"
        />
         
        25
      </span>
    </p>
  </div>
</section>
`;

exports[`<Post /> matches the trending post snapshot 1`] = `
<section
  className="Post__trending"
>
  <img
    alt=""
    className="Post__trending__image"
    src="https://source.unsplash.com/140x140/?astronomy&sig=efgh2i"
  />
  <div
    className="Post__trending__content"
  >
    <h2
      className="Post__trending__flair"
    >
      Astronomy
    </h2>
    <a
      className="Post__trending__link"
      href="/posts/astronomy/efgh2i"
      onClick={[Function]}
    >
      <h3
        className="Post__trending__title"
        id="efgh2i"
        onClick={[Function]}
      >
        Where Can the Solar Eclipse be Viewed...
      </h3>
    </a>
  </div>
</section>
`;
